articles/testing-documentation-snippets/article.md

```diff
@@ -0,0 +1,173 @@
+[snippet-code]
+
+```
+
+The tag **[snippet-code]** in the template will be replaced by the snippet and will form a full context file which can be executed.
+
+![schema2](./images/schema2.png)
+
+
+At this point, we have many source files, in many languages. First version of this new documentation system can support snippets test for Javascript, Java, C++ and Go. Some languages are interpreted, others compiled. All of them need their dedicated toolchain !
+
+![use docker](./images/use-docker.jpg)
+
+First thing I do is to create a DockerFile for each language implemented with all the toolchain to compile/interpret generated file. Also, we decided to add a linter in each container for each language to be sure to have a correct syntax and minimize source of errors (you can see an exemple of a dockerfile [here](https://hub.docker.com/r/kuzzleio/documentation-v2/~/dockerfile/)).
+
+At this point, we had some docker containers able to run code for their dedicated language and things seem to work perfectly!
+
+![schema3](./images/schema3.png)
+
+
+## Implementing test system
+
+![doge](./images/doge.png)
+
+Next task was to implement the test system properly speaking. How to test all of these languages with one test definition? Finally that's simple, all languages can write in the standard output. Javascript has his `console.log`, C++ his `std::cout`, Go his `fmt.Println` etc, etc ... With the help of my teammates we decided to use **Expect** to check if the tests are passed or not.
+
+> *From wikipedia :* Expect is an automation and non-regression testing tool written by Don Libes as an extension to the Tcl scripting language to test interactive applications like telnet, ftp, passwd, fsck, rlogin, ssh, or others.
+
+Running tests is simple, we give to expect a command to execute, also a string expected and it can do assertion on the result of the command.
+
+![schema expect](./images/schema-expect.png)
+
+
+For writing these tests, we opted for a YAML file in front of the snippets files. It simple to write, simple to read and simple to parse. They have to describe the test, give an expected string and the template we want to use. We also imagined a system of hooks (before and after) like in every test framework to execute some script for cleaning or other operations.
+
+```yaml
+name: kuzzle-connect
+description: Connects the SDK to Kuzzle
+hooks:
+  before: 
+  after: 
+template: default
+expected: Successfully connected
+```
+
+I found a cool Node.js package on NPM named [Nexpect](https://github.com/nodejitsu/nexpect). Then, I wrote a Node.js script for running tests for each language. For an easy implementation of every languages I make an abstract Tester class and I implement a children class for each language with lint and run command. In this way if we want to add another language for our SDKs it's very easy to just create a class and extend it to the mother tester class and only override commands to be executed for this new language.
```

Comment:
" I make an abstract Tester" :point_right: "I made an abstract `Tester` class and I implemented a children class for each language with the `lint` and `run` command"

"In this way if we want to add another language for our SDKs it's very easy to just create a class and extend it to the mother tester class and only override commands to be executed for this new language." :point_down: 
"In this way if we want to add another language for our SDKs it's very easy to create a new `Tester` by just extending the mother tester class and only override the commands to be executed for this new language."